lib/input/reader/amqp_1.go

```diff
@@ -231,15 +249,26 @@ func (a *AMQP1) ReadWithContext(ctx context.Context) (types.Message, AsyncAckFn,
 
 	msg.Append(part)
 
-	var done chan struct{}
+	var renewDoneChannel chan struct{}
 	if a.conf.AzureRenewLock {
-		done = a.startRenewJob(amqpMsg)
+		renewDoneChannel = a.startRenewJob(amqpMsg)
 	}
 
+	a.m.RLock()
+	a.pendingMsgs[pendingMsgIndex(amqpMsg)] = &amqp1PendingMsg{
+		msg: msg,
+		renewDoneChannel: renewDoneChannel,
+	}
+	a.m.RUnlock()
+
 	return msg, func(ctx context.Context, res types.Response) error {
-		if done != nil {
-			close(done)
-			done = nil
+		msg, ok := a.popPendingMessage(amqpMsg)
+		if !ok {
+			return nil
+		}
+
+		if msg.renewDoneChannel != nil {
+			close(msg.renewDoneChannel)
```

Comment:
Should we close the channel in the `popPendingMessage` instead ?
Then you could also call this `popPenddingMessage` in the cleanup loop on connection?

Reply:
`popPendingMessage` locks the map so we want it to be as quick as possible, also in the cleanup loop, we already hold the lock so we cannot use `popPendingMessage` there.